### ISSUE TITLE
Wrong date and (or) time in all RSS feeds

### ISSUE THREAD

**Issue by TitovLab** [dokuwiki/dokuwiki#2960]
In the release 2017-02-19f "Frusterick Manners" the dates were displayed correctly in RSS feeds, but after updating to 2018-04-22b "Greebo" there are errors:
Types ATOM 0.3, ATOM 1, RSS 1 always show 01.01.2020 date for feeds.
Types RSS 0.9 and RSS 2 show wrong feed time (add time zone).

For ATOM 0.3, ATOM 1, RSS 1 "Greebo" show time as `2020-01-15T10:17:14+0300`, "Frusterick Manners" showed `2020-01-15T10:17:14+03:00`
For RSS 0.9 and RSS 2 "Greebo" show time as `Wed, 15 Jan 2020 10:17:14 +0000`, "Frusterick Manners" showed `Wed, 15 Jan 2020 10:17:14 +0300`

Apparently everything is in the iso8601 and rfc822 functions in the file /vendor/openpsa/universalfeedcreator/lib/Element/FeedDate.php.
In "Frusterick Manners" these functions were in /inc/feedcreator.class.php and looked a bit different. I returned them as before:
```
--- FeedDate.php        2020-01-15 10:34:43.340813900 +0300
+++ FeedDate.php        2020-01-15 10:52:32.487813900 +0300
@@ -97,8 +97,8 @@
     public function rfc822()
     {
         //return gmdate("r",$this->unix);
-        $date = gmdate("D, d M Y H:i:s O", $this->unix);
-
+        $date = gmdate("D, d M Y H:i:s ", $this->unix);
+        if (TIME_ZONE!="") $date .= " ".str_replace(":","",TIME_ZONE);
         return $date;
     }

@@ -110,11 +110,10 @@
     public function iso8601()
     {
         $date = gmdate("Y-m-d\TH:i:sO", $this->unix);
-        $date = substr($date, 0, 22).':'.substr($date, -2);
         if (TIME_ZONE != "") {
-            $date = str_replace("+00:00", TIME_ZONE, $date);
+            $date = str_replace("+0000", TIME_ZONE, $date);
         }
-
+        $date = substr($date, 0, 22).':'.substr($date, -2);
         return $date;
     }
```
After that, the errors disappeared, but there is a feeling that this one is still wrong, because the time zone can be not only a plus, but also a minus.

PHP 7.3.13, timezone set in php.ini
Microsoft Outlook 2016 as RSS client.

**Comment by splitbrain** [comment]
I guess this should be reported at https://github.com/flack/UniversalFeedCreator

I wonder if it's actually related to flack/UniversalFeedCreator#6

**Comment by fjf2002** [comment]
@splitbrain:

1. https://packagist.org/packages/openpsa/universalfeedcreator links to https://github.com/flack/UniversalFeedCreator, so: YES, they should be related.

2. I've opened a ticket there, see https://github.com/flack/UniversalFeedCreator/issues/14 .

3. Nevertheless consider the following concerning the dokuwiki project: UniversalFeedCreator internally works with a class FeedDate, which in turn internally uses unixtime integers.

So let's take a look at the following dokuwiki code:

https://github.com/dokuwiki/dokuwiki/blob/91a25860ebdf15b20aa33b98d798cd0c04851206/feed.php#L244

You convert a unixtime integer to an RFC 2822 string, which in turn will get converted back to unixtime by the FeedDate class.
I advise to assign plain unixtime to `$item->date`, which not only avoids back-and-forth conversion, but also circumvents the UniversalFeedCreator timezone problem.

**Comment by lprobsth** [comment]
I think this was fixed upstream.

I have a test install of dokuwiki with the newest master branch where UniversalFeedCreator was installed with composer. The version of UniversalFeedCreator there is 1.8.5 (although version 1.8.6 is already out). The timestamps in the feed are correct there.

It seems that the changes in UniversalFeedCreator to `FeedDate` were introduced in version 1.8.5.

I have also manually installed version 1.8.6 of UniversalFeedCreator in a Dokuwiki Jack Jackrum installation and that seems to work.
